src/lean_spec/subspecs/containers/attestation/attestation.py

```diff
@@ -82,5 +87,103 @@ class SignedAggregatedAttestations(Container):
+    """
+    Extract all validator indices encoded in aggregation bits.
+
+    Returns the list of all validators who participated in the aggregation,
+    sorted by validator index.
+
+    Args:
+        bits: Aggregation bitlist with participating validators.
+
+    Returns:
+        List of validator indices, sorted in ascending order.
+    """
+    validator_indices = [Uint64(index) for index, bit in enumerate(bits) if bool(bit)]
+    if not validator_indices:
+        raise AssertionError("Aggregated attestation must reference at least one validator")
+    return validator_indices
+
+
+def aggregated_attestations_to_plain(
+    aggregated: AggregatedAttestation,
+) -> list[Attestation]:
+    """
+    Convert aggregated attestation to a list of plain Attestation containers.
+
+    Extracts all participating validator indices from the aggregation bits
+    and creates individual Attestation objects for each validator.
+
+    Args:
+        aggregated: Aggregated attestation with one or more participating validators.
+
+    Returns:
+        List of plain attestations, one per participating validator.
+    """
+    validator_indices = aggregation_bits_to_validator_indices(aggregated.aggregation_bits)
+    return [
+        Attestation(validator_id=validator_id, data=aggregated.data)
+        for validator_id in validator_indices
+    ]
+
+
+def attestation_to_aggregated(attestation: Attestation) -> AggregatedAttestation:
+    """Convert a plain Attestation into the aggregated representation."""
+    validator_index = int(attestation.validator_id)
+    bits = [False] * (validator_index + 1)
+    bits[validator_index] = True
+    return AggregatedAttestation(
+        aggregation_bits=AggregationBits(data=bits),
+        data=attestation.data,
+    )
+
+
+def aggregate_attestations_by_data(
+    attestations: list[Attestation],
+) -> list[AggregatedAttestation]:
+    """
+    Aggregate attestations with common attestation data.
+
+    Groups attestations by their AttestationData and creates one AggregatedAttestation
+    per unique data, with all participating validator bits set.
+
+    Args:
+        attestations: List of attestations to aggregate.
+
+    Returns:
+        List of aggregated attestations with proper bit aggregation.
+    """
+    # Group validator IDs by attestation data (avoids intermediate objects)
+    data_to_validator_ids: dict[AttestationData, list[int]] = defaultdict(list)
+
+    for attestation in attestations:
+        data_to_validator_ids[attestation.data].append(int(attestation.validator_id))
+
+    # Create aggregated attestations with all relevant bits set
+    result: list[AggregatedAttestation] = []
+
+    for data, validator_ids in data_to_validator_ids.items():
+        # Find the maximum validator index to determine bitlist size
+        max_validator_id = max(validator_ids)
+
+        # Create bitlist with all participating validators set to True
+        bits = [False] * (max_validator_id + 1)
```

Comment:
this is wrong, ~~this way all the bits will be flipped to true~~, you have to use max registry limit as the max validator count

Reply:
ahh ok, this is sort of an optimization where the list doesn't go high if the the max validator id in that list is low
add a comment to this effect: that on an average case for packing sparse attestations this size could be lower than full list size